tokio-trace/src/lib.rs

```diff
@@ -0,0 +1,760 @@
+                if !event.is_disabled() {
+                    let mut keys = callsite.metadata().fields().into_iter();
+                    let msg_key = keys.next()
+                        .expect("event metadata should define a key for the message");
+                    $(
+                        let key = keys.next()
+                            .expect(concat!("metadata should define a key for '", stringify!($k), "'"));
+                        event!(@ record: event, $k, &key, $($val)*);
+                    )*
+                }
+            }
+            event
+        }
+    });
+    (target: $target:expr, $lvl:expr, $($arg:tt)+ ) => (
+        event!(target: $target, $lvl, { }, $($arg)+)
+    );
+    ( $lvl:expr, { $( $k:ident $( = $val:expr )* ),* }, $($arg:tt)+ ) => (
+        event!(target: module_path!(), $lvl, { $($k $( = $val)* ),* }, $($arg)+)
+    );
+    ( $lvl:expr, $( $k:ident $( = $val:expr )* ),* ) => (
+        event!(target: module_path!(), $lvl, { $($k $( = $val)* ),* })
+    );
+    ( $lvl:expr, $($arg:tt)+ ) => (
+        event!(target: module_path!(), $lvl, { }, $($arg)+)
+    );
+    (@ record: $ev:expr, $k:expr, $i:expr, $val:expr) => (
+        $ev.record($i, &$val);
+    );
+    (@ record: $ev:expr, $k:expr, $i:expr,) => (
+        // skip
+    );
+}
+
+/// Constructs an event at the trace level.
+///
+/// When both a message and fields are included, curly braces (`{` and `}`) are
+/// used to delimit the list of fields from the format string for the message.
+///
+/// # Examples
+///
+/// ```rust
+/// # #[macro_use]
+/// # extern crate tokio_trace;
+/// # use std::time::SystemTime;
+/// # #[derive(Debug, Copy, Clone)] struct Position { x: f32, y: f32 }
+/// # impl Position {
+/// # const ORIGIN: Self = Self { x: 0.0, y: 0.0 };
+/// # fn dist(&self, other: Position) -> f32 {
+/// #    let x = (other.x - self.x).exp2(); let y = (self.y - other.y).exp2();
+/// #    (x + y).sqrt()
+/// # }
+/// # }
+/// # fn main() {
+/// use tokio_trace::field;
+///
+/// let pos = Position { x: 3.234, y: -1.223 };
+/// let origin_dist = pos.dist(Position::ORIGIN);
+///
+/// trace!(position = field::debug(pos), origin_dist = field::debug(origin_dist));
+/// trace!(target: "app_events",
+///         { pos = field::debug(pos) },
+///         "x is {} and y is {}",
+///        if pos.x >= 0.0 { "positive" } else { "negative" },
+///        if pos.y >= 0.0 { "positive" } else { "negative" });
+/// # }
+/// ```
+#[macro_export]
+macro_rules! trace {
+    (target: $target:expr, { $( $k:ident $( = $val:expr )* ),* }, $($arg:tt)+ ) => (
+        event!(target: $target, $crate::Level::TRACE, { $($k $( = $val)* ),* }, $($arg)+)
+    );
+    (target: $target:expr, $( $k:ident $( = $val:expr )* ),* ) => (
+        event!(target: $target, $crate::Level::TRACE, { $($k $( = $val)* ),* })
+    );
+    (target: $target:expr, $($arg:tt)+ ) => (
+        drop(event!(target: $target, $crate::Level::TRACE, {}, $($arg)+));
```

Comment:
Why does this drop the event, but the others don't? If there's good reason, perhaps a tiny comment for Future Us?

Reply:
Added a comment in c37623000e53dfdf5f4d3e4aa832b4da12f6a2cf.